Add newlib retargetable locking implementation and tests

stephanosio opened the pull request:
This series adds the newlib retargetable locking interface function implementations, as well as the tests to verify that the retargetable locking interface is functional and used by the internal lock functions.

```
This commit adds the newlib retargetable locking interface function
implementations in order to make newlib functions thread safe.

The newlib retargetable locking interface is internally called by the
standard C library functions provided by newlib to synchronise access
to the internal shared resources.

By default, the retargetable locking interface functions defined within
the newlib library are no-op. When multi-threading is enabled (i.e.
`CONFIG_MULTITHREADING=y`), the Zephyr-side retargetable locking
interface implementations override the default newlib implementation
and provide locking mechanism.

The retargetable locking interface may be called with either a static
(`__lock__...`) or a dynamic lock.

The static locks are statically allocated and initialised immediately
after kernel initialisation by `newlib_locks_prepare`.

The dynamic locks are allocated and de-allocated through the
`__retargetable_lock_init[_recursive]` and
`__retarget_lock_close_[recurisve]` functions as necessary by the
newlib functions. These locks are allocated in the newlib heap using
the `malloc` function when userspace is not enabled -- this is safe
because the internal multi-threaded malloc lock implementations
(`__malloc_lock` and `__malloc_unlock`) call the retargetable locking
interface with a static lock (`__lock__malloc_recursive_mutex`). When
userspace is enabled, the dynamic locks are allocated and freed through
`k_object_alloc` and `k_object_release`.

Note that the lock implementations used here are `k_mutex` and `k_sem`
instead of `sys_mutex` and `sys_sem` because the Zephyr kernel does not
currently support dynamic allocation of the latter. These locks should
be updated to use `sys_mutex` and `sys_sem` when the Zephyr becomes
capable of dynamically allocating them in the future.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

Fixes the "Retarget Locking" portion of #21519.

NOTE1: The reentrancy-related fixes will be addressed in an upcoming (separate) PR.
NOTE2: The tests will fail because the Zephyr SDK newlib is currently compiled with `--disable-newlib-multithread`. To be merged after https://github.com/zephyrproject-rtos/sdk-ng/pull/344 is mainlined in the CI.

Fixes #21519.

stephanosio committed:
tests: newlib: thread_safety: Fix test function scope

This commit adds the `static` keyword to the test functions that are
not intended to be globally available.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
lib: newlib: Add retargetable locking implementation

This commit adds the newlib retargetable locking interface function
implementations in order to make newlib functions thread safe.

The newlib retargetable locking interface is internally called by the
standard C library functions provided by newlib to synchronise access
to the internal shared resources.

By default, the retargetable locking interface functions defined within
the newlib library are no-op. When multi-threading is enabled (i.e.
`CONFIG_MULTITHREADING=y`), the Zephyr-side retargetable locking
interface implementations override the default newlib implementation
and provide locking mechanism.

The retargetable locking interface may be called with either a static
(`__lock__...`) or a dynamic lock.

The static locks are statically allocated and initialised immediately
after kernel initialisation by `newlib_locks_prepare`.

The dynamic locks are allocated and de-allocated through the
`__retargetable_lock_init[_recursive]` and
`__retarget_lock_close_[recurisve]` functions as necessary by the
newlib functions. These locks are allocated in the newlib heap using
the `malloc` function when userspace is not enabled -- this is safe
because the internal multi-threaded malloc lock implementations
(`__malloc_lock` and `__malloc_unlock`) call the retargetable locking
interface with a static lock (`__lock__malloc_recursive_mutex`). When
userspace is enabled, the dynamic locks are allocated and freed through
`k_object_alloc` and `k_object_release`.

Note that the lock implementations used here are `k_mutex` and `k_sem`
instead of `sys_mutex` and `sys_sem` because the Zephyr kernel does not
currently support dynamic allocation of the latter. These locks should
be updated to use `sys_mutex` and `sys_sem` when the Zephyr becomes
capable of dynamically allocating them in the future.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: newlib: thread_safety: Add internal lock and userspace tests

This commit adds the tests for the newlib retargetable locking
interface, as well as the tests for the internal lock functions that
are supposed to internally invoke the retargetable locking interface.

All of these tests must pass when the toolchain newlib is compiled with
the `retargetable-locking` and `multithread` options, which are
required to ensure that the newlib is thread-safe, enabled. If the
toolchain newlib is compiled with either of these options disabled,
this test will fail.

This commit also adds the userspace testcases to ensure that the newlib
is thread-safe in the user mode.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented on lib/libc/newlib/libc-hooks.c:
```diff
@@ -298,17 +299,130 @@ void *_sbrk(intptr_t count)
+{
+	__ASSERT_NO_MSG(lock != NULL);
+
+	/* Allocate mutex object */
+	*lock = malloc(sizeof(struct sys_mutex));
+	__ASSERT(*lock != NULL, "recursive lock allocation failed");
+
+	sys_mutex_init((struct sys_mutex *)*lock);
```
So, unlike `sys_sem`, `sys_mutex` does not provide the userspace implementation (i.e. futex) and `sys_mutex_init` is, for some reason, no-op simply relying on the build system to provide the kernel-side object only for the statically allocated `sys_sem`s.

This dynamically allocated `sys_mutex` should not work when `CONFIG_USERSPACE=y` because there is no corresponding `k_mutex` kernel object:

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/lib/os/mutex.c#L33-L42

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/lib/os/mutex.c#L12-L22

I would say the fact that `sys_mutex_init` is a no-op, when there is no futex, is a bug.

@andyross Any comments on this?

stephanosio replied:
After a more in-depth analysis on the current Zephyr userspace implementation, even the `k_futex` used internally by `sys_mutex` will have problems when there is a contention because the `k_futex` relies on the `z_futex_data` kernel object.

Ideally, it would be possible to dynamically allocate these objects through `k_object_alloc`, but the current implementation prohibits that -- the `Lives in user memory` rationale given here is incorrect because these "kernel" objects are supposed to be the protected kernel-side objects for the corresponding user objects.
https://github.com/zephyrproject-rtos/zephyr/blob/d6c9d40ee062f7dd1f8edae461e21c8fe88824db/kernel/userspace.c#L347-L354

FWIW, I think the current userspace `sys_sem` and `sys_mutex` implementations require some major rework to provide kind of abstraction that it was supposed to provide. Alternatively, we could simply declare that dynamic allocation of these objects are simply not allowed given the nature of Zephyr "RT"OS, but that sounds like a major limitation for more advanced applications.

@andyross Your insight would be greatly appreciated.

stephanosio replied:
Related issues:
* #25497 `k_futex` unnecessarily relies on the kernel object management
* #15138 `sys_mutex` does not use `k_futex` for userspace

stephanosio commented:
UPDATE:
* Changed newlib retargetable locking to use `k_mutex` and `k_sem` instead of `sys_mutex` and `sys_sem` since the latter [do not support dynamic allocation due to the limitation of the current Zephyr kernel implementation](https://github.com/zephyrproject-rtos/zephyr/pull/36201#discussion_r650628137).
* Added userspace test coverage (now tests both non-userspace and userspace builds).

stephanosio commented:
Marking it as "ready for review" since the PR itself is ready. Please note that the CI failure will be gone as soon as the SDK 0.13.0 is mainlined.

stephanosio commented:
SDK 0.13.0 has been mainlined. This PR is ready for final review and merging.

Please note that the checkpatch failure is wrong and should be ignored.

carlescufi reviewed (approved):
This looks good to me, with my limited knowledge of newlib's retargetable locking system.

galak reviewed (changes requested):
We need to update the `TOOLCHAIN_ZEPHYR_MINIMUM_REQUIRED_VERSION` in the cmakefiles

stephanosio committed:
cmake: Set minimum required Zephyr SDK version to 0.13

In order to use the newlib retagetable locking interface (for thread
safety), which requires the newlib multi-threading feature to be
enabled, the Zephyr SDK 0.13.0 or above must be used.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
> We need to update the TOOLCHAIN_ZEPHYR_MINIMUM_REQUIRED_VERSION in the cmakefiles

@galak Done

cfriedt commented:
CI says there is a whitespace issue. I'm not sure I see what the problem is though. @galak - could that be a false positive?

stephanosio commented:
> CI says there is a whitespace issue. I'm not sure I see what the problem is though. @galak - could that be a false positive?

@cfriedt checkpatch has a bug in which it fails to recognise variable declarations with volatile pointers. Yes, it is a false positive.

galak commented:
@cfriedt I went ahead and merged this.

cfriedt commented:
> @cfriedt I went ahead and merged this.

Thanks - I don't seem to have merge rights to manually merge PR's in case of false positives, but I'm also kind of ok with that.
